### ISSUE TITLE
JIT model with `mul+atan+sgn` will access illegal memory on cuda when computing gradient

### ISSUE THREAD

**Issue by YangChenyuan** [pytorch/pytorch#86529]
### 🐛 Describe the bug

JIT model with `mul+atan+sgn` will access illegal memory on cuda when computing gradient

```py
import torch

class M(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, input):
        input = torch.mul(input, torch.tensor(13, dtype=torch.float32, device='cuda'))
        input = torch.atan(input)
        fn_res = torch.sgn(input, )
        return fn_res

fn = M().to('cuda')
torch.random.manual_seed(18986)
inp = torch.empty([5, 5, 5], dtype=torch.float64)
inp.uniform_(-1, 127)
inp = inp.to('cuda')

from torch.autograd.functional import jacobian

jit_fn = torch.jit.script(fn)
jit_fn(inp.clone().requires_grad_())
print(jacobian(jit_fn, inp.clone().requires_grad_()))
```

### Versions

```
PyTorch version: 1.13.0.dev20221001
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.2.0-19ubuntu1) 11.2.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.13 (main, Aug 25 2022, 23:26:10)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.19.5-051905-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.7.99
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 515.65.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.3
[pip3] torch==1.13.0.dev20221001
[pip3] torchaudio==0.13.0.dev20221001
[pip3] torchvision==0.14.0.dev20221001
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0            py39h7f8727e_0  
[conda] mkl_fft                   1.3.1            py39hd3c417c_0  
[conda] mkl_random                1.2.2            py39h51133e4_0  
[conda] numpy                     1.23.3                   pypi_0    pypi
[conda] numpy-base                1.23.1           py39ha15fc14_0  
[conda] pytorch                   1.13.0.dev20221001 py3.9_cuda11.7_cudnn8.5.0_0    pytorch-nightly
[conda] pytorch-cuda              11.7                 h67b0de4_0    pytorch-nightly
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torch                     1.13.0.dev20220923+cu116          pypi_0    pypi
[conda] torchaudio                0.13.0.dev20220923+cu116          pypi_0    pypi
[conda] torchvision               0.14.0.dev20220923+cu116          pypi_0    pypi
```

**Comment by YangChenyuan** [comment]
It seems that this bug will also appear on CPU

```py
import torch

class M(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, input):
        input = torch.mul(input, torch.tensor(13, dtype=torch.float32, device='cpu'))
        input = torch.atan(input)
        fn_res = torch.sgn(input, )
        return fn_res

fn = M().to('cpu')
torch.random.manual_seed(18986)
inp = torch.empty([5, 5, 5], dtype=torch.float64)
inp.uniform_(-1, 127)
inp = inp.to('cpu')

from torch.autograd.functional import jacobian

print(fn(inp))
jit_fn = torch.jit.script(fn)

jit_fn(inp.clone().requires_grad_())
print(jacobian(jit_fn, inp.clone().requires_grad_()))
```

```
tensor([[[ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.],
         [ 1., -1.,  1.,  1.,  1.]],

        [[ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.]],

        [[ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.]],

        [[ 1.,  1.,  1.,  1.,  1.],
         [ 1., -1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.]],

        [[ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.],
         [ 1.,  1.,  1.,  1.,  1.]]], dtype=torch.float64)

segmentation fault (core dumped)
```

**Comment by jjsjann123** [comment]
I'm trying to figure out what's wrong here....

So the issue repros on both NNC & nvfuser. I suspected that it's coming from autodiff. But if I disable fusion, it actually passed with autodiff giving similar fw/bw graph.... hmmm pretty mysterious... I'll keep digging this one.

**Comment by jjsjann123** [comment]
This is the forward optimized graph:
```
Optimized Graph: 
graph(%self : __torch__.M,
      %input.1 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=1, device=cuda:0)):
  %27 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=1, device=cuda:0), %28 : bool = prim::RequiresGradCheck[types=[Tensor(requires_grad=1)]](%input.1)
  %29 : Tensor = prim::If(%28)
    block0():
      %14 : Tensor = prim::DifferentiableGraph_0(%27)
      -> (%14)
    block1():
      %38 : Function = prim::Constant[name="fallback_function", fallback=1]()
      %39 : (Tensor) = prim::CallFunction(%38, %input.1) 
      %40 : Tensor = prim::TupleUnpack(%39)
      -> (%40)
  %fn_res.1 : Tensor = aten::sgn(%29) # /jiej/playground/repro_86529.py:12:17
  return (%fn_res.1)
with prim::DifferentiableGraph_0 = graph(%input.1 : Tensor(requires_grad=0)): 
  %9 : Tensor(requires_grad=0) = prim::Constant[value={13}]()
  %input.14 : Tensor(requires_grad=0) = aten::mul(%input.1, %9) # /jiej/playground/repro_86529.py:10:16
  %51 : int[] = aten::size(%input.1) # <string>:3:44 
  %52 : int[] = aten::size(%input.14) # <string>:3:55
  %53 : int[]? = aten::_size_if_not_equal(%51, %52) # <string>:3:19
  %54 : int[] = aten::size(%9) # <string>:3:93 
  %56 : int[]? = aten::_size_if_not_equal(%54, %52) # <string>:3:68
  %57 : (int[]?, int[]?) = prim::TupleConstruct(%53, %56)
  %self_size.4 : int[]?, %other_size.2 : int[]? = prim::TupleUnpack(%57)
  %input.10 : Tensor = aten::atan(%input.14) # /jiej/playground/repro_86529.py:11:16
  return (%input.10, %9, %self_size.4, %input.14)
```

This is the optimized graph from the differentiable graph:
```
Optimized Graph:
graph(%input.1 : Tensor):
  %2 : Float(requires_grad=0, device=cuda:0) = prim::Constant[value={13}]()
  %6 : int[] = aten::size(%input.1) # <string>:3:44
  %22 : bool = prim::CudaFusionGuard[types=[Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0)]](%input.1)
  %20 : Tensor, %21 : Tensor = prim::If(%22)
    block0():
      %input.19 : Tensor, %input.17 : Tensor = prim::CudaFusionGroup_0[cache_id=0](%input.1)
      -> (%input.19, %input.17)
    block1():
      %28 : Function = prim::Constant[name="fallback_function", fallback=1]()
      %29 : (Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0), Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0)) = prim::CallFunction(%28, %input.1)
      %30 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0), %31 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0) = prim::TupleUnpack(%29)
      -> (%30, %31)
  %19 : int[] = aten::size(%21)
  %9 : int[]? = aten::_size_if_not_equal(%6, %19) # <string>:3:19
  return (%20, %2, %9, %21)
with prim::CudaFusionGroup_0 = graph(%2 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0)):
  %3 : float = prim::Constant[value=13.]()
  %input.14 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0) = aten::mul(%2, %3) # /jiej/playground/repro_86529.py:10:16
  %input.10 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0) = aten::atan(%input.14) # /jiej/playground/repro_86529.py:11:16
  return (%input.10, %input.14)
```

Arguments from the fused kernel (sorry about double printing outputs, I need to patch that):
```
Arguments for kernel1:
Inputs:
tensor dtype: double sizes: (5, 5, 5, ) stride: (25, 5, 1, ) pointer: 0x7fa918200a00
tensor dtype: double sizes: (5, 5, 5, ) stride: (25, 5, 1, ) pointer: 0x7fa918200e00
tensor dtype: double sizes: (5, 5, 5, ) stride: (25, 5, 1, ) pointer: 0x7fa918201600
Outputs:
  Double [5, 5, 5] (strides = [25, 5, 1], address = 0x7fa918200e00)
  Double [5, 5, 5] (strides = [25, 5, 1], address = 0x7fa918201600)
```

This is the backward graph:
```
Optimized Graph:
graph(%0 : Tensor,
      %1 : Tensor,
      %2 : Tensor,
      %self_size.5 : int[]?,
      %4 : Tensor):
  %167 : NoneType = prim::Constant()
  %168 : bool = aten::__is__(%self_size.5, %167)
  %169 : bool = prim::AutogradAllNonZero(%0, %2, %4)
  %170 : bool = prim::AutogradAllZero(%1)
  %171 : bool[] = prim::ListConstruct(%168, %169, %170)
  %172 : bool = aten::all(%171)
  %55 : (Tensor) = prim::If(%172)
    block0():
      %182 : bool = prim::CudaFusionGuard[types=[Float(requires_grad=0, device=cuda:0), Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0), Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0)]](%2, %0, %4)
      %181 : Tensor = prim::If(%182)
        block0():
          %173 : Tensor = prim::CudaFusionGroup_0[cache_id=1](%2, %0, %4)
          -> (%173)
        block1():
          %189 : Function = prim::Constant[name="fallback_function", fallback=1]()
          %190 : (Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0)) = prim::CallFunction(%189, %2, %0, %4)
          %191 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0) = prim::TupleUnpack(%190)
          -> (%191)
      %110 : (Tensor) = prim::TupleConstruct(%181)
      -> (%110)
    block1():
      %192 : Function = prim::Constant[name="fallback_function", fallback=1]()
      %193 : ((Tensor)) = prim::CallFunction(%192, %0, %1, %2, %self_size.5, %4)
      %194 : (Tensor) = prim::TupleUnpack(%193)
      -> (%194)
  return (%55)
with prim::CudaFusionGroup_0 = graph(%1 : Float(requires_grad=0, device=cuda:0),
      %3 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0),
      %9 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0)):
  %7 : int = prim::Constant[value=1]() # <string>:201:52
  %10 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0) = aten::mul(%9, %9) # <string>:201:38
  %8 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0) = aten::add(%10, %7, %7) # <string>:201:38
  %5 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0) = aten::div(%3, %8) # <string>:201:23
  %2 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0) = aten::mul(%5, %1) # <string>:10:29
  return (%2)
```

Similarly arguments for fusion we have:
```
Arguments for kernel2:
Inputs:
tensor dtype: float sizes: () stride: () pointer: 0x7fa918200800
tensor dtype: double sizes: (5, 5, 5, ) stride: (25, 5, 1, ) pointer: 0
tensor dtype: double sizes: (5, 5, 5, ) stride: (25, 5, 1, ) pointer: 0x7fa918201600
tensor dtype: double sizes: (5, 5, 5, ) stride: (25, 5, 1, ) pointer: 0x7fa918200e00
Outputs:
  Double [5, 5, 5] (strides = [25, 5, 1], address = 0x7fa918200e00)
```

Notice that the second tensor input to the fusion has a `0` pointer, if you trace the graph, this should be the `%input.1` on the forward differentiable graph. So I think there's something wrong with autodiff...
But strangely this only happens when we have fusion enabled.... maybe there's something in the optimization pass that messes up the lifetime of the tensor.

**Comment by YangChenyuan** [comment]
Hi @jjsjann123, thanks for your help! Sorry that I am not familiar with the optimization with autodiff.

I am curious why there are 4 inputs for fusion `kernel2`. Is `kernel2` the following function?

```
with prim::CudaFusionGroup_0 = graph(%1 : Float(requires_grad=0, device=cuda:0),
      %3 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0),
      %9 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0)):
```

**Comment by jjsjann123** [comment]
> Hi @jjsjann123, thanks for your help! Sorry that I am not familiar with the optimization with autodiff.
> 
> I am curious why there are 4 inputs for fusion `kernel2`. Is `kernel2` the following function?
> 
> ```
> with prim::CudaFusionGroup_0 = graph(%1 : Float(requires_grad=0, device=cuda:0),
>       %3 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0),
>       %9 : Double(5, 5, 5, strides=[25, 5, 1], requires_grad=0, device=cuda:0)):
> ```

Sorry, that's a bug in our print. The last input is actually the outputs. :stuck_out_tongue: So just discard that....

**Comment by YangChenyuan** [comment]
I see 🤣

Is the tensor with the null pointer at the start of the backward graph or just before the fusion call?

I think it would be hard to debug this bug since it composes of different parts. Maybe we need help from a developer who is familiar with the autodiff :)
